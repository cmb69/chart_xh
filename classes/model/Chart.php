<?php

/**
 * Copyright (c) Christoph M. Becker
 *
 * This file is part of Chart_XH.
 *
 * Chart_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Chart_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Chart_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Chart\Model;

use DOMDocument;
use DOMElement;
use DOMNode;
use Plib\Document2;
use Plib\DocumentStore2;

final class Chart implements Document2
{
    private string $name;
    private string $caption;
    /** @var list<string> */
    private array $labels = [];
    /** @var list<Dataset> */
    private array $datasets = [];

    public static function new(string $key): self
    {
        return new self(basename($key, ".xml"), "");
    }

    public static function fromString(string $contents, string $key): ?self
    {
        if ($contents === "") {
            return null;
        }
        $doc = new DOMDocument("1.0", "UTF-8");
        if (!$doc->loadXML($contents)) {
            return null;
        }
        if (!$doc->relaxNGValidate(__DIR__ . "/../../chart.rng")) {
            return null;
        }
        assert($doc->documentElement instanceof DOMElement);
        $chart = $doc->documentElement;
        $that = new self(basename($key, ".xml"), $chart->getAttribute("caption"));
        foreach ($chart->childNodes as $childNode) {
            assert($childNode instanceof DOMNode);
            if ($childNode->nodeName === "label") {
                assert($childNode instanceof DOMElement);
                $that->labels[] = (string) $childNode->nodeValue;
            } elseif ($childNode->nodeName === "dataset") {
                assert($childNode instanceof DOMElement);
                $that->datasets[] = Dataset::fromXml($childNode);
            }
        }
        return $that;
    }

    public static function create(string $name, DocumentStore2 $store): self
    {
        $that = $store->create("$name.xml", self::class);
        assert($that instanceof self);
        return $that;
    }

    public static function read(string $name, DocumentStore2 $store): ?self
    {
        return $store->read("$name.xml", self::class);
    }

    public static function update(string $name, DocumentStore2 $store): ?self
    {
        return $store->update("$name.xml", self::class);
    }

    public function __construct(string $name, string $caption)
    {
        $this->name = $name;
        $this->caption = $caption;
    }

    public function name(): string
    {
        return $this->name;
    }

    public function caption(): string
    {

        return $this->caption;
    }

    /** @return list<string> */
    public function labels(): array
    {
        return $this->labels;
    }

    /** @return list<Dataset> */
    public function datasets(): array
    {
        return $this->datasets;
    }

    public function setCaption(string $caption): void
    {
        $this->caption = $caption;
    }

    public function addLabel(string $label): void
    {
        $this->labels[] = $label;
    }

    public function addDataset(string $label, string $color): Dataset
    {
        return $this->datasets[] = new Dataset($label, $color);
    }

    public function toString(): string
    {
        $doc = new DOMDocument('1.0', 'UTF-8');
        $chart = $doc->createElement('chart');
        $chart->setAttribute("caption", $this->caption);
        $doc->appendChild($chart);
        foreach ($this->labels as $label) {
            $elt = $doc->createElement("label");
            $elt->nodeValue = $label;
            $chart->appendChild($elt);
        }
        foreach ($this->datasets as $dataset) {
            $chart->appendChild($dataset->toXml($doc));
        }
        if (!$doc->relaxNGValidate(__DIR__ . "/../../chart.rng")) {
            return "";
        }
        $doc->formatOutput = true;
        return (string) $doc->saveXML();
    }
}
