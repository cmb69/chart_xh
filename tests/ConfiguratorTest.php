<?php

namespace Chart;

use ApprovalTests\Approvals;
use Chart\Model\Chart;
use PHPUnit\Framework\TestCase;
use Plib\DocumentStore2 as DocumentStore;

/** @medium */
class ConfiguratorTest extends TestCase
{
    private DocumentStore $store;

    public function setUp(): void
    {
        $this->store = new DocumentStore("./examples/");
    }

    private function sut(): Configurator
    {
        return new Configurator();
    }

    public function testConfiguresChart(): void
    {
        $chart = Chart::read("voting", $this->store);
        Approvals::verifyAsJson($this->sut()->configure($chart));
    }

    public function testConfiguresTransposedChart(): void
    {
        $chart = Chart::read("voting", $this->store);
        $chart->setTransposed(true);
        Approvals::verifyAsJson($this->sut()->configure($chart));
    }
}
