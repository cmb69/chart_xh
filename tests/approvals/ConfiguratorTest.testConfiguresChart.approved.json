{
    "data": {
        "datasets": [
            {
                "backgroundColor": "#ff0000",
                "borderColor": "#ff0000",
                "data": [
                    20.5,
                    25.7,
                    16.4
                ],
                "label": "SPD"
            },
            {
                "backgroundColor": "#000000",
                "borderColor": "#000000",
                "data": [
                    33,
                    24.2,
                    28.6
                ],
                "label": "CDU\/CSU"
            },
            {
                "backgroundColor": "#00ff00",
                "borderColor": "#00ff00",
                "data": [
                    8.9,
                    14.7,
                    11.6
                ],
                "label": "Gr\u00fcne"
            },
            {
                "backgroundColor": "#ffff00",
                "borderColor": "#ffff00",
                "data": [
                    10.7,
                    11.4,
                    4.3
                ],
                "label": "FDP"
            },
            {
                "backgroundColor": "#0000ff",
                "borderColor": "#0000ff",
                "data": [
                    12.6,
                    10.4,
                    20.8
                ],
                "label": "AfD"
            },
            {
                "backgroundColor": "#ff00ff",
                "borderColor": "#ff00ff",
                "data": [
                    9.2,
                    4.9,
                    8.8
                ],
                "label": "Die Linke"
            },
            {
                "backgroundColor": "#7f7f7f",
                "borderColor": "#7f7f7f",
                "data": [
                    5,
                    8.7,
                    9.6
                ],
                "label": "Sonstige"
            }
        ],
        "labels": [
            "2017",
            "2021",
            "2025"
        ]
    },
    "options": {
        "spanGaps": true
    },
    "type": "bar"
}